File: 66.Plus-One.go
// https://leetcode.com/problems/plus-one/
//
// algorithms
// Easy (40.37%)
// Total Accepted:    327,582
// Total Submissions: 812,114
// beats 100.0% of golang submissions

package leetcode

func plusOne(digits []int) []int {
	length := len(digits)

	if digits[length-1] != 9 {
		digits[length-1]++
		return digits
	}

	digits[length-1] = 0
	idx := length - 2
	for ; idx >= 0; idx-- {
		if digits[idx] != 9 {
			digits[idx]++
			return digits
		} else {
			digits[idx] = 0
		}
	}

	digits = append([]int{1}, digits...)

	return digits
}

File: 66.Plus-One.py
# https://leetcode.com/problems/plus-one/
#
# algorithms
# Easy (41.61%)
# Total Accepted:    434,239
# Total Submissions: 1,043,613


class Solution(object):
    def plusOne(self, digits):
        """
        :type digits: List[int]
        :rtype: List[int]
        """
        length = len(digits)
        carry = 1

        for i in xrange(length - 1, -1, -1):
            tmp = digits[i] + 1
            if tmp < 10:
                digits[i] = tmp
                carry = 0
                break

            else:
                carry = 1
                digits[i] = tmp - 10

        if carry == 1:
            digits.insert(0, 1)

        return digits
